vault/diagnose/tls.go

```diff
@@ -41,12 +69,36 @@ func TLSFileChecks(certFilePath, keyFilePath string) error {
 	}
 	cert.Leaf = x509Cert
 
-	// TODO: Check root as well via l.TLSClientCAFile
+	if !checkRoot {
+		// Check that certificate isn't expired and is of correct usage type
+		if _, err = cert.Leaf.Verify(x509.VerifyOptions{
+			KeyUsages: []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		}); err != nil {
+			return fmt.Errorf("failed to verify certificate: " + err.Error())
+		}
+	} else {
+		if rootFilePath == "" {
+			return fmt.Errorf("TLS Root CA file not specified, but TLS chain verification is required")
+		}
+		caPool := x509.NewCertPool()
+		data, err := ioutil.ReadFile(rootFilePath)
+		if err != nil {
+			return fmt.Errorf("failed to read tls_client_ca_file: %s", err.Error())
+		}
+
+		if !caPool.AppendCertsFromPEM(data) {
+			return fmt.Errorf("failed to parse CA certificate in tls_client_ca_file")
+		}
+		// Check that certificate isn't expired and is of correct usage type
+		if _, err = cert.Leaf.Verify(x509.VerifyOptions{
+			Roots:         caPool,
+			Intermediates: x509.NewCertPool(),
+			KeyUsages:     []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		}); err != nil {
+			return fmt.Errorf("failed to verify certificate: " + err.Error())
```

Comment:
you're using a formatted error here without formatting. i think the same `%w` verb bit could be good here too

Reply:
`fmt.Errorf` is the only Error call in the `fmt` package, so I have to use it. But to your point, I'm happy to change all of these error messages to use `%w` instead of `%s` with `err.Error()` :).

Reply:
yeah, i guess i just thought it looked a little strange with a `"" + <var>` thing instead of a `"<some verb>", <var>` thing, but great!